content/blog/15-making-rust-attractive-for-writing-gtk-applications.md

```diff
@@ -0,0 +1,301 @@
+        type ParentType = gtk::Widget;
+        type Instance = subclass::simple::InstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
+
+        glib_object_subclass!();
+
+        fn new() -> Self {
+            Self()
+        }
+    }
+
+    impl ObjectImpl for SimpleWidget {
+        glib_object_impl!();
+    }
+    impl WidgetImpl for SimpleWidget {}
+}
+glib::wrapper! {
+    pub struct SimpleWidget(ObjectSubclass<imp::SimpleWidget>)
+        @extends gtk::Widget;
+}
+```
+
+The same code nowadays
+```rust
+mod imp {
+    #[derive(Default)]
+    pub struct SimpleWidget;
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for SimpleWidget {
+        const NAME: &'static str = "SimpleWidget";
+        type ParentType = gtk::Widget;
+    }
+
+    impl ObjectImpl for SimpleWidget {}
+    impl WidgetImpl for SimpleWidget {}
+}
+glib::wrapper! {
+    pub struct SimpleWidget(ObjectSubclass<imp::SimpleWidget>)
+        @extends gtk::Widget;
+}
+```
+
+Of course, the overall code is still a bit too verbose particularly when you have to define `GObject` properties. Although, people have been experimenting with writing a [derive macro](https://github.com/gtk-rs/gtk-rs-core/pull/494) to simplify the properties declaration part as well as a [`gobject::class!`](https://github.com/gtk-rs/gtk-rs-core/issues/540) macro for generating most of the remaining boilerplate code.
+
+
+### Composite templates
+
+In short, it allows you to make your custom `GtkWidget` subclass uses `XML` for providing the widget structure and so allows splitting the UI code from it logic. The UI part can be either inlined in the code or written in a separate file and unless you are using `gresources`, you can opt-in for runtime validation of the XML content using the `xml_validation` feature.
+
+```rust
+mod imp {
+    #[derive(Default, gtk::CompositeTemplate)]
+    #[template(string = r#"
+    <interface>
+      <template class="SimpleWidget" parent="GtkWidget">
+        <child>
+          <object class="GtkLabel" id="label">
+            <property name="label">foobar</property>
+          </object>
+        </child>
+      </template>
+    </interface>
+    "#)]
+    pub struct SimpleWidget {
+        #[template_child]
+        pub label: TemplateChild<gtk::Label>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for SimpleWidget {
+        const NAME: &'static str = "SimpleWidget";
+        type ParentType = gtk::Widget;
+
+        fn class_init(klass: &mut Self::Class) {
+            klass.bind_template();
+        }
+
+        fn instance_init(obj: &gtk::glib::subclass::InitializingObject<Self>) {
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for SimpleWidget {
+        fn dispose(&self) {
+            // since gtk 4.8
+            self.obj().dispose_template(Self::Type);
+            // before gtk 4.8, for each direct child widget
+            // self.label.unparent();
```

Comment:
I would put what most people do here, which would be:

```rust
while let Some(child) = self.obj().first_child() {
    child.unparent();
}
```

Reply:
I'd also add a note that `dispose_template()` only works for templates - if you're making a non-template custom widget, you'd need this while-let loop still.